Basic e2e tests failing randomly

I tried to investigate the `basic e2e` tests and why they're failing randomly as at this point having them pass is sort of a gamble.
After running tests and investigating I found two ways they fail, both of these should be investigated and underlying issues fixed.

- This is the most common failure:
```
....
  --- FAIL: TestConsumerConnectsToProvider/Consumers_Connect_to_provider (52.81s)
        connection_test.go:359: 
            	Error Trace:	connection_test.go:359
            	            				connection_test.go:167
            	            				asm_amd64.s:1357
            	Error:      	Condition never satisfied
            	Test:       	TestConsumerConnectsToProvider/Consumers_Connect_to_provider
            	Messages:   	Consumer reported spending 1191184987523 hermes says <nil>. Service type openvpn. Hermes url http://hermes:8889/api/v2, consumer ID 0x5e1b877e949334334b89406382e5ae420bd48575
    --- FAIL: TestConsumerConnectsToProvider/Validate_provider_earnings (5.04s)
        connection_test.go:611: 
            	Error Trace:	connection_test.go:611
            	            				connection_test.go:178
            	Error:      	Condition never satisfied
            	Test:       	TestConsumerConnectsToProvider/Validate_provider_earnings
        connection_test.go:622: 
            	Error Trace:	connection_test.go:622
            	            				connection_test.go:178
            	Error:      	Not equal: 
            	            	expected: 16430787589190
            	            	actual  : 15239602601668
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-(*big.Int)(16430787589190)
            	            	+(*big.Int)(15239602601668)
            	            	 
            	Test:       	TestConsumerConnectsToProvider/Validate_provider_earnings
            	Messages:   	consumers reported spend 16430787589190, providers earnings 15239602601668
        connection_test.go:623: 
            	Error Trace:	connection_test.go:623
            	            				connection_test.go:178
            	Error:      	Not equal: 
            	            	expected: 16430787589190
            	            	actual  : 15239602601668
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-(*big.Int)(16430787589190)
            	            	+(*big.Int)(15239602601668)
            	            	 
            	Test:       	TestConsumerConnectsToProvider/Validate_provider_earnings
            	Messages:   	consumers reported spend 16430787589190, providers earnings 15239602601668
    --- FAIL: TestConsumerConnectsToProvider/Provider_settlement_flow (0.89s)
        connection_test.go:217: 
            	Error Trace:	connection_test.go:217
            	Error:      	Not equal: 
            	            	expected: 15239602601668
            	            	actual  : 16430787589190
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-(*big.Int)(15239602601668)
            	            	+(*big.Int)(16430787589190)
            	            	 
            	Test:       	TestConsumerConnectsToProvider/Provider_settlement_flow
        connection_test.go:233: 
            	Error Trace:	connection_test.go:233
            	Error:      	Should be true
            	Test:       	TestConsumerConnectsToProvider/Provider_settlement_flow
```

- This is less common but also happens:
```
  === RUN   TestConsumerConnectsToProvider
=== RUN   TestConsumerConnectsToProvider/Provider_has_a_registered_identity
=== RUN   TestConsumerConnectsToProvider/Consumer_Creates_And_Registers_Identity
{"level":"info","time":"2020-12-30T11:35:52Z","message":"Created new identity: 0x5d08785dc3cd7ae0e45afde0daac379f4d57c084"}
{"level":"info","time":"2020-12-30T11:35:52Z","message":"Created new identity: 0xaca6818d0d4c2082f5d5877ceb6364a12a6969ea"}
{"level":"info","time":"2020-12-30T11:35:52Z","message":"Created new identity: 0x860a44f7c5dc1ad27c9800ad1130dcdd8873748f"}
{"level":"info","time":"2020-12-30T11:35:52Z","message":"Created new identity: 0xbaf3e1548082c5811f5f3bd8d05bf36caa578186"}
=== RUN   TestConsumerConnectsToProvider/Consumers_Connect_to_provider
panic: runtime error: index out of range [0] with length 0

goroutine 256 [running]:
github.com/mysteriumnetwork/node/e2e.consumerPicksProposal(0xc0001e5b00, 0xc0005681c0, 0x18218af, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	go/src/github.com/mysteriumnetwork/node/e2e/connection_test.go:509 +0x2be
github.com/mysteriumnetwork/node/e2e.TestConsumerConnectsToProvider.func3.1(0xc00056b500, 0xc0001e5b00, 0x2c9ba00)
	go/src/github.com/mysteriumnetwork/node/e2e/connection_test.go:163 +0xba
created by github.com/mysteriumnetwork/node/e2e.TestConsumerConnectsToProvider.func3
	go/src/github.com/mysteriumnetwork/node/e2e/connection_test.go:161 +0xad
```

--- a/e2e/connection_test.go
+++ b/e2e/connection_test.go
@@ -40,6 +40,7 @@ import (
 	tequilapi_client "github.com/mysteriumnetwork/node/tequilapi/client"
 	"github.com/mysteriumnetwork/node/tequilapi/contract"
 	"github.com/mysteriumnetwork/payments/bindings"
+	"github.com/mysteriumnetwork/payments/crypto"
 )
 
 var (
@@ -53,6 +54,7 @@ var (
 	registryAddress             = "0xbe180c8CA53F280C7BE8669596fF7939d933AA10"
 	channelImplementation       = "0x599d43715DF3070f83355D9D90AE62c159E62A75"
 	addressForTopups            = "0xa29fb77b25181df094908b027821a7492ca4245b"
+	hundredthThou               = float64(1) / float64(100000)
 )
 
 var ethClient *ethclient.Client
@@ -354,6 +356,7 @@ func TestConsumerConnectsToProvider(t *testing.T) {
 }
 
 func recheckBalancesWithHermes(t *testing.T, consumerID string, consumerSpending *big.Int, serviceType, hermesURL string) {
+	var testSuccess bool
 	var lastHermes *big.Int
 	assert.Eventually(t, func() bool {
 		hermesCaller := pingpong.NewHermesCaller(requests.NewHTTPClient("0.0.0.0", time.Second), hermesURL)
@@ -361,8 +364,38 @@ func recheckBalancesWithHermes(t *testing.T, consumerID string, consumerSpending
 		assert.NoError(t, err)
 		promised := hermesData.LatestPromise.Amount
 		lastHermes = promised
-		return promised.Cmp(consumerSpending) == 0
-	}, time.Second*20, time.Millisecond*300, fmt.Sprintf("Consumer reported spending %v hermes says %v. Service type %v. Hermes url %v, consumer ID %v", consumerSpending, lastHermes, serviceType, hermesURL, consumerID))
+
+		// Author: vkuznecovas
+		// Due to the async nature of the payment system, a situation might occur where a consumers reported spending is larger than the actual amount that reaches hermes.
+		// This happens in the following flow:
+		// 1) Session is established.
+		// 2) Payments occur normally for some time.
+		// 3) Consumer received yet another invoice.
+		// 4) Consumer starts calculating the amount to promise.
+		// 5) Session is killed as operation 4) is happening.
+		// 6) The promise is incremented but never issued as the session is aborted. This can happen due to promise not being sent, or provider not listening for further promises on the given topic.
+		// 7) Hermes is not aware of the last promise, therefore the consumer and hermes reportings are different.
+		// I do not believe this will cause issues in reality, as such situations occur in e2e test regularly due to the rapid exchange of promises.
+		// Under normal circumstances, such occurences should be VERY, VERY rare and the amount of myst involved is rather small. They should have no impact on the payment system as a whole.
+		// Therefore, for these tests to be stable, the following solution is proposed:
+		// Make sure that the hermes and consumer reported spendings differ by no more than 1/100000 of a myst.
+		absDiffFloat := getDiffFloat(consumerSpending, promised)
+		res := absDiffFloat < hundredthThou
+		if res {
+			testSuccess = true
+		}
+		return res
+	}, time.Second*20, time.Millisecond*300)
+
+	if !testSuccess {
+		fmt.Printf("Consumer reported spending %v hermes says %v. Service type %v. Hermes url %v, consumer ID %v", consumerSpending, lastHermes, serviceType, hermesURL, consumerID)
+	}
+}
+
+func getDiffFloat(a, b *big.Int) float64 {
+	diff := new(big.Int).Sub(a, b)
+	absoluteDiff := new(big.Int).Abs(diff)
+	return crypto.BigMystToFloat(absoluteDiff)
 }
 
 func identityCreateFlow(t *testing.T, tequilapi *tequilapi_client.Client, idPassphrase string) string {
@@ -602,8 +635,14 @@ func providerEarnedTokens(t *testing.T, tequilapi *tequilapi_client.Client, id s
 	providerStatus, err := tequilapi.Identity(id)
 	assert.NoError(t, err)
 	assert.True(t, providerStatus.Balance.Cmp(new(big.Int)) == 0)
-	assert.Equal(t, earningsExpected, providerStatus.Earnings, fmt.Sprintf("consumers reported spend %v, providers earnings %v", earningsExpected, providerStatus.Earnings))
-	assert.Equal(t, earningsExpected, providerStatus.EarningsTotal, fmt.Sprintf("consumers reported spend %v, providers earnings %v", earningsExpected, providerStatus.Earnings))
+
+	// For reasoning behind these, see the comment in recheckBalancesWithHermes
+	actualEarnings := getDiffFloat(earningsExpected, providerStatus.Earnings)
+	assert.True(t, actualEarnings < hundredthThou)
+
+	actualEarnings = getDiffFloat(earningsExpected, providerStatus.EarningsTotal)
+	assert.True(t, actualEarnings < hundredthThou)
+
 	assert.True(t, providerStatus.Earnings.Cmp(big.NewInt(500)) == 1, "earnings should be at least 500 but is %d", providerStatus.Earnings)
 	return providerStatus.Earnings
 }

